### DISCUSSION TITLE
What is the `target` for macOS bundles in GitHub Actions pipeline?

### DISCUSSION THREAD

**Discussion by elanzini** [tauri-apps/tauri#6090] [Q&A] [answered]
From the [macOs Bundle](https://tauri.app/v1/guides/building/macos/) docs, it says that `By default, the CLI builds a binary targeting your machine's architecture`. I am building the installers using GitHub Actions with the specification found on the [Code Signing macOS](https://tauri.app/v1/guides/distribution/sign-macos/) docs. The platform is set to `macos-latest`.

In this case, what is the target being used? 
How do you produce two different installers with targets `aarch64-apple-darwin` and `x86_64-apple-darwin`?

**Comment by FabianLars** [comment, marked as answer]
github's runners are x86_64. To build 2 installers for both you need to change the build matrix from the [example](https://tauri.app/v1/guides/building/cross-platform#configuration) to include the target, something like this: https://github.com/FabianLars/mw-toolbox/blob/6a59339cbcb3ec6127f3656bce543b1c392f5b05/.github/workflows/temp.yml#L23-L27 (this file has some unnecessary stuff because i use it for testing tauri-action PRs).
If you use tauri-action then you also need to use it from dev `tauri-apps/tauri-action@dev`. Lastly you just have to pipe the target arg into the action: https://github.com/FabianLars/mw-toolbox/blob/6a59339cbcb3ec6127f3656bce543b1c392f5b05/.github/workflows/temp.yml#L88

**Reply by elanzini** [reply]
Thanks @FabianLars. I see in one of the configs you are not supplying any `--target`. Does that break the pipeline?
e.g. in my case for window I am fine to build on `windows-latest` without target.

**Reply by FabianLars** [reply]
nope it doesn't break it. This is effectively the same as not providing it at all.

**Reply by elanzini** [reply]
That's great. One last thing. Do you know why the macOS builds are so expensive on GitHub actions in terms of minutes? Every release is taking more than 200 minutes when checking the usage in the Billing page.

**Reply by FabianLars** [reply]
Because Apple requires that macOS runs on original Apple hardware which is stupidly expensive in comparison. To compensate that build minutes on macOS runners are multiplied by 10. :/

btw github actions is completely free for public repos in case that's a possibility for your project.

**Reply by ahkohd** [reply]
> That's great. One last thing. Do you know why the macOS builds are so expensive on GitHub actions in terms of minutes? Every release is taking more than 200 minutes when checking the usage in the Billing page.

you can setup a self-hosted GitHub runner, this way you can use your idle macOS hardware at your home (or even the mac you're using for development) as a runner for your GitHub actions

**Reply by elanzini** [reply]
> you can use a self-hosted GitHub runner, this way you can use your idle macOS hardware as a runner for your GitHub actions.

Interesting - do you have any docs about it?

**Reply by ahkohd** [reply]
> > That's great. One last thing. Do you know why the macOS builds are so expensive on GitHub actions in terms of minutes? Every release is taking more than 200 minutes when checking the usage in the Billing page.
> 
> you can set up a self-hosted GitHub runner; this way, you can use your idle macOS hardware at your home (or even the mac you're using for development) as a runner for your GitHub actions

When I learnt this trick, I saved myself a lot of money. At one point, I was paying up to $50/mo for Github action that compiles my Tauri macOS project. The most annoying thing is that when singing the app, the process hangs the Github runner; however, the runner keeps running but doing nothing, accumulating cost. It was as if I was paying Github actions for doing nothing 😥

**Reply by ahkohd** [reply]
> > you can use a self-hosted GitHub runner, this way you can use your idle macOS hardware as a runner for your GitHub actions.
> 
> Interesting - do you have any docs about it?

it's really easy to setup:
https://docs.github.com/en/actions/hosting-your-own-runners

**Reply by elanzini** [reply]
@FabianLars I am seeing this error in the pipeline: 
```
Error failed to build app: Target aarch64-apple-darwin is not installed (installed targets: x86_64-apple-darwin). Please run `rustup target add aarch64-apple-darwin`.
```
I am running with `tauri-apps/tauri-action@dev`. I checked your config and I see that you have this:
```yml
    - uses: dtolnay/rust-toolchain@stable
      with:
        targets: aarch64-apple-darwin
```
Is this what I am missing?

**Reply by FabianLars** [reply]
whoops, yes it is. Sorry.

This will install the host's default toolchain&target and additionally the aarch64-apple-darwin target. The way i specified it it will install the target on linux and windows too, so in case you don't want to waste a second you probably can wrap that in some kind of if clause like similar to the libwebkitgtk package installation

**Reply by elanzini** [reply]
@FabianLars Is the `dev` version of `tauri-apps/tauri-action` different compared to `v0`? 
Does it expect a release draft to be created already? Pipeline fails again with this error.

```
Looking for a draft release with tag app-v0.0.1...
Couldn't find release with tag app-v0.0.1. Creating one.
Error: Resource not accessible by integration
```

**Reply by FabianLars** [reply]
> Is the dev version of tauri-apps/tauri-action different compared to v0?

Not really, no.

> Does it expect a release draft to be created already?

Nope, the first 2 lines of the log are normal.

Try manually specifying the permissions like this: https://github.com/tauri-apps/tauri-action/pull/286/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R57

**Reply by elanzini** [reply]
@FabianLars That worked - many thanks as always ☺️
